rts/Lua/LuaSyncedCtrl.cpp

```diff
@@ -1754,29 +1757,107 @@ int LuaSyncedCtrl::TransferUnit(lua_State* L)
 	if (unit == nullptr)
 		return 0;
 
-	const int newTeam = luaL_checkint(L, 2);
-	if (!teamHandler.IsValidTeam(newTeam))
+	const int newTeamID = luaL_checkint(L, 2);
+	if (!teamHandler.IsValidTeam(newTeamID))
 		return 0;
 
-	const CTeam* team = teamHandler.Team(newTeam);
-	if (team == nullptr)
+	CTeam* newTeam = teamHandler.Team(newTeamID);
+	if (newTeam == nullptr)
 		return 0;
 
 	bool given = true;
 	if (FullCtrl(L) && lua_isboolean(L, 3))
 		given = lua_toboolean(L, 3);
 
+	bool adjustUnitLimit = false;
+	if (lua_isboolean(L, 4))
+		adjustUnitLimit = lua_toboolean(L, 4);
+
+	CTeam* oldTeam = teamHandler.Team(unit->team);
+	if (oldTeam == nullptr) {
+		return 0;
+	}
+
 	if (inTransferUnit >= MAX_CMD_RECURSION_DEPTH)
 		luaL_error(L, "TransferUnit() recursion is not permitted, max depth: %d", MAX_CMD_RECURSION_DEPTH);
 
 	++ inTransferUnit;
 	ASSERT_SYNCED(unit->id);
-	ASSERT_SYNCED((int)newTeam);
+	ASSERT_SYNCED((int)oldTeam->teamNum);
+	ASSERT_SYNCED((int)newTeamID);
 	ASSERT_SYNCED(given);
-	unit->ChangeTeam(newTeam, given ? CUnit::ChangeGiven
+	ASSERT_SYNCED(adjustUnitLimit);
+	if (adjustUnitLimit) {
+		newTeam->maxUnits++;
+		oldTeam->maxUnits--;
+	}
+	bool successfulTransfer = unit->ChangeTeam(newTeamID, given ? CUnit::ChangeGiven
 	                                : CUnit::ChangeCaptured);
+	if (adjustUnitLimit && !successfulTransfer) {
+		newTeam->maxUnits--;
+		oldTeam->maxUnits++;
+	}
 	-- inTransferUnit;
-	return 0;
+	
+	lua_pushboolean(L, successfulTransfer);
+	return 1;
+}
+
+
+/***
+ * @function Spring.TransferUnitLimit
+ * @number oldTeamID
+ * @number newTeamID
+ * @number transferAmnt
+ * @treturn bool successfulTransfer
+ */
+int LuaSyncedCtrl::TransferUnitLimit(lua_State* L)
+{
+	CheckAllowGameChanges(L);
+	
+	const int oldTeamID = luaL_checkint(L, 1);
+	if (!teamHandler.IsValidTeam(oldTeamID))
+		return 0;
+
+	const int newTeamID = luaL_checkint(L, 2);
+	if (!teamHandler.IsValidTeam(newTeamID))
+		return 0;
+
+	CTeam* oldTeam = teamHandler.Team(oldTeamID);
+	if (oldTeam == nullptr)
+		return 0;
+
+	CTeam* newTeam = teamHandler.Team(newTeamID);
+	if (newTeam == nullptr)
+		return 0;
+
+	const int transferAmnt = luaL_checkint(L, 3);
+	
+	// Only allow oldTeam to give a positive amount to newTeam
+	if (transferAmnt <= 0)
+		return 0;
+
+	// Make sure transferAmnt is in range
+	if (oldTeam->numUnits > oldTeam->maxUnits - transferAmnt // transferAmnt too big for current numUnits
```

Comment:
I think I would like everything from below this point to be a teamHandler function such that this Lua func would just do
```cpp
book ok = teamHandler->TransferUnitLimit(oldTeam, newTeam, amount);
```
since there are other places where the limit is transferred (for example handling killed teams) that would ideally be handled the same way.